pkg/raft/term_cache.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package raft
+
+// TermCache represents a compressed representation of the suffix of a raftLog.
+// It is used to determine the term of a raft entry at a specific index.
+//
+// The TermCache conforms to raft safety properties. It also provides the
+// following guarantees:
+//
+//  1. cache[i].Index < entries[i+1].Index,
+//  2. cache[i].term < entries[i+1].Term.
+//  3. cache[0] is always populated.
+//  4. Except for the first entry cache[0], all entries are term flips.
+//
+// Which we can use the above invariants to determine the term of a
+// raft entry if given an index that is greater than cache[0].index.
+//
+// Users of this struct can assume the invariants hold true.
```

Comment:
We should also state the expectation of this type. It observes all appends, truncations, and (re-)initializations of the `raftLog`.